app/controllers/foreman_ansible/api/v2/hostgroups_controller_extensions.rb

```diff
@@ -45,10 +47,63 @@ def multiple_play_roles
 
             render_message @result
           end
+
+          api :GET, '/hostgroups/:id/list_ansible_roles',
+              N_('Lists assigned Ansible roles')
+          param :id, :identifier, :required => true
+
+          def list_ansible_roles
+            find_resource
+
+            @result = {
+              :roles => @hostgroup.all_ansible_roles
+            }
+
+            render_message @result
+          end
+
+          api :POST, '/hostgroups/:id/ansible_roles',
+              N_('Assigns Ansible roles to a host group')
+          param :id, :identifier, :required => true
+          param :roles, Array, :required => true
+
+          def ansible_roles
+            find_resource
+
+            @hostgroup.ansible_roles = @roles - \
```

Comment:
It's fine to have the roles show up several times in the hierarchy, like:

hostgroup A - roles: "epel", "git"
hostgroup B - roles: "epel", "note"

Did you remove the inherited roles for any reason?